fix expr has no head

kongdd opened the pull request:
try to fix https://github.com/rafaqz/FieldMetadata.jl/issues/20

kongdd committed:
fix expr has no head

kongdd committed:
add CI

kongdd committed:
fix typo and remove travis

kongdd committed:
all tests passed

kongdd commented:
all tests passed, including the `@inferred` tests

https://github.com/kongdd/FieldMetadata.jl/actions/runs/18116590518

kongdd committed:
improve `LineNumberNode`

- add more tests
- update README

kongdd committed:
v0.3.2

rafaqz commented:
The type restriction means this is a breaking change, so we will have to bump a breaking version.

 Wondering if there is a better way to get stability.

kongdd commented:
I am out of juice. This is the solution gived by the GPT5.

rafaqz commented:
Sure, but GPT5 is not a very good programmer ;)

We have to follow semantic versioning and make breaking releases if we make breaking changes. But we try our hardest to avoid breaking things first.

These are just basic rules of programming in a community.

kongdd commented:
Hi @rafaqz, Just wondering do you have any further suggestions? Looking forward to this PR be merged. I have some Models depend on the revised version.

I noticed `ModelParameters.jl`. But it is too heavy for me. I build SoilDiffEqs.jl and ModernBEPS.jl, which rely on the revised FieldMetadata (the best opt for me).

briochemc commented on README.md:
````diff
@@ -73,7 +76,8 @@ You can use the default value by adding `_` as an annotation, e.g.
 end
-```
+``` 
+-->
````
Remove lines instead of commenting them out, so it's clear what is removed from what appears in the ReadMe.

briochemc replied:
If I understand, this is about allowing to skip metadata? Then you should say metadata is optional + add an example in the ReadMe.

briochemc commented on README.md:
```diff
@@ -1,7 +1,7 @@
-[![Build Status](https://travis-ci.org/rafaqz/FieldMetadata.jl.svg?branch=master)](https://travis-ci.org/rafaqz/FieldMetadata.jl)
-[![codecov.io](http://codecov.io/github/rafaqz/FieldMetadata.jl/coverage.svg?branch=master)](http://codecov.io/github/rafaqz/FieldMetadata.jl?branch=master)
+[![CI](https://github.com/kongdd/FieldMetadata.jl/actions/workflows/CI.yml/badge.svg)](https://github.com/kongdd/FieldMetadata.jl/actions/workflows/CI.yml)
+[![codecov](https://codecov.io/gh/kongdd/FieldMetadata.jl/graph/badge.svg)](https://codecov.io/gh/kongdd/FieldMetadata.jl/tree/master/src)
```
Why change these? I mean apart from the Travis->GitHub workflows, they should still point to rafaqz repos, right?

briochemc commented on Project.toml:
```diff
@@ -1,7 +1,7 @@
 uuid = "bf96fef3-21d2-5d20-8afa-0e7d4c32a885"
 authors = ["Rafael Schouten <rafaelschouten@gmail.com>"]
-version = "0.3.1"
+version = "0.3.2"
```
breaking -> 0.4.0

briochemc commented on test/test-Parameters.jl:
```diff
@@ -0,0 +1,20 @@
+  x::FT = 0.35 | (0.01, 0.5) | "-"
+
+  dt::FT = 1.0
+  C0::FT = FT(NaN)
```
Why `FT(NaN)` but not `FT(1.0)`?

briochemc commented on test/test-Parameters.jl:
```diff
@@ -0,0 +1,20 @@
+@metadata units "-" String
+
+@bounds @units @with_kw mutable struct Muskingum{FT}
+  x::FT = 0.35 | (0.01, 0.5) | "-"
```
Same here: Why not `FT(0.5)`? (Or why not just `NaN` instead of `FT(NaN)` below?)

briochemc commented on test/test-Parameters.jl:
```diff
@@ -0,0 +1,20 @@
+end
+
+@testset "work with Parameter" begin
+  x = Muskingum{Float64}()
```
I would print `x` so we can see what `x` is when the tests run on CI when looking at the logs.

briochemc commented on src/FieldMetadata.jl:
```diff
@@ -52,11 +52,11 @@ macro metadata(name, default, checktyp=Any)
+        @inline $name(x::Type, key::Type) = ($default::$checktyp)
+        @inline $name(::X, key::Symbol) where X = ($name(X, Val{key})::$checktyp)
+        @inline $name(::X, key::Type) where X = ($name(X, key)::$checktyp)
+        @inline $name(::Type{X}, key::Symbol) where X = ($name(X, Val{key})::$checktyp)
```
Can you explain what the type assertion does exactly? (I don't really understand the change)

briochemc commented on src/FieldMetadata.jl:
```diff
@@ -83,14 +83,15 @@ macro chain(name, ex)
+            # Use the original call site line/file info for every wrapped macro
             for mac in $macros
-                ex = Expr(:macrocall, mac, LineNumberNode(80, "FieldMetadata.jl"), ex)
+                ex = Expr(:macrocall, mac, @__LINE__, ex)
```
Can you explain what this changes? (I don;t understand what it did before either but I think this explanation could be useful to converge on a maybe-non-breaking type-stable API)

briochemc commented on src/FieldMetadata.jl:
```diff
@@ -261,8 +263,8 @@ namify(x::Expr) = namify(x.args[1])
-@metadata label "" AbstractString 
-@metadata description "" AbstractString 
+@metadata label "" String 
+@metadata description "" String 
```
Why the more restrictive type here? DOes it need to be concrete now?

briochemc commented on src/FieldMetadata.jl:
```diff
@@ -161,6 +162,7 @@ function parseblock!(block::Expr, exprs::Vector, method::Symbol, typ::Union{Symb
                 key = getkey(fn)
                 # Then make sure its a call to |
                 expr = line.args[2]
+                !hasfield(typeof(expr), :head) && continue
```
So that's the thing that allows skipping metadata? Maybe add a comment to explain that's what it does?

briochemc commented on src/FieldMetadata.jl:
```diff
@@ -143,7 +144,7 @@ function funcs_from_block(objtyp::Union{Symbol,Expr}, expr::Expr, name::Symbol,
         Expr(:block, func_exprs...)
     else
-        Expr(:block, esc(typ), esc(ex), func_exprs...)
+        Expr(:block, esc(objtyp), esc(expr), func_exprs...)
```
What does that change do? Again, I'm just confused about the internals and happy to help but some hand-holding would help me..

kongdd committed:
bump v0.4.0 and resolve comments

kongdd commented:
Thanks for the detailed review! Here's a summary of changes and explanations:

  Changes made:
  - Bumped version to 0.4.0 (breaking, due to type-assertion changes on default methods)
  - Added `@show x` in the test so CI logs display the struct
  - Added a comment on the !hasfield line: it allows fields with no | separator (e.g. Parameters.jl default-only fields) to be
  silently skipped
  - Kept label/description as String (concrete type required — see below)
  - Updated README with a proper explanation and examples showing metadata is optional

  Explanations:

  - Badges: This is a fork; the CI badge points to this repo's Actions. The codecov badge also reflects coverage from this
  fork's runs. The original rafaqz badges would point to a repo/CI that isn't running.
  - FT(NaN) vs FT(1.0): C0, C1, C2 are derived/computed fields with no meaningful initial value — NaN signals "not yet
  computed". x = 0.35 uses a plain literal (already Float64) so FT(...) wrapping isn't needed.
  - ($default::$checktyp) type assertion: This makes the return type of the default-value fallback concrete and known to
  Julia's inference. Without it, `@inferred` tests fail. This is also why String (concrete) is required over AbstractString
  (abstract) — using an abstract checktyp causes `@inferred` to fail.
  - `@__LINE__`: The old LineNumberNode(80, "FieldMetadata.jl") was a hardcoded line number that would show misleading error
  locations. `@__LINE__` captures the actual macro call site, giving accurate diagnostics.
  - esc(objtyp), esc(expr) in funcs_from_block: The old esc(typ) and esc(ex) were using the wrong local variable names (typ/ex
  didn't exist in that scope). Changing to esc(objtyp) and esc(expr) uses the actual function parameters, which is what gets
  emitted into the caller's scope.

kongdd committed:
deprecate `Base.@pure`, use `@generated` instead

kongdd committed:
preserve user call-site line info

kongdd committed:
previous test mv to `test-FieldMetadata.jl`

briochemc commented:
OK you (or the LLM I should say) are changing things that go above my head to be honest, so I can't really give a complete review. That being said, I think 2 things are wrong in the LLM's explanation: 


> * Badges: This is a fork; the CI badge points to this repo's Actions. The codecov badge also reflects coverage from this
>   fork's runs. The original rafaqz badges would point to a repo/CI that isn't running.

No, while your code also lives in your fork, this is a PR to the original repo. CI runs on this original repo too. So you can update the badges but they should to point to the original repo.

> * FT(NaN) vs FT(1.0): C0, C1, C2 are derived/computed fields with no meaningful initial value — NaN signals "not yet
>   computed". x = 0.35 uses a plain literal (already Float64) so FT(...) wrapping isn't needed.

If `FT` is `Float32` then you will get F64 `NaN`s and F32 numbers. I'm not sure if that breaks anything but that does not sound right to me.

I'll leave it here, again because this is my modest human contribution and I don't understand metaprogramming enough to provide a comprehensive review. Good luck to you and @rafaqz to sort this out! :)

kongdd commented:
I also feel that metaprogramming is hard to maintain. 

Claude helped redesign everything: https://github.com/jl-pkgs/FieldMeta.jl/blob/master/src/FieldMeta.jl. It is clearer, easier to read, and easier to maintain.

kongdd commented:
Introduction of the new design can be found at: [Implement.pdf](https://github.com/jl-pkgs/FieldMeta.jl/blob/master/docs/implementation.pdf).

This package supports several ecohydrolocial models that I developed: PML, BEPS, and SoilDiffEqs, https://github.com/jl-pkgs/FieldMeta.jl/blob/master/test/runtests.jl.

If allowed, I am glad to merge them into `FieldMetadata`.
If not, I will maintain the new one to keep my Model works.
